libs/vkd3d/command.c

```diff
@@ -2538,176 +2544,186 @@ static bool d3d12_command_list_update_graphics_pipeline(struct d3d12_command_lis
-            return false;
+            ERR("Unhandled descriptor type %d.\n", binding->type);
     }
 
+    return false;
+}
+
+static bool vkd3d_descriptor_updates_resize_arrays(struct vkd3d_descriptor_updates *updates,
+        unsigned int descriptor_count)
+{
+    /* This should grow over time to the point where no further allocations are necessary */
+    if (!vkd3d_array_reserve((void **)&updates->descriptors, &updates->descriptors_size,
+            descriptor_count, sizeof(*updates->descriptors)))
+        return false;
+
+    if (!vkd3d_array_reserve((void **)&updates->descriptor_writes, &updates->descriptor_writes_size,
+            descriptor_count, sizeof(*updates->descriptor_writes)))
+        return false;
+
     return true;
 }
 
-static void d3d12_command_list_update_descriptor_table(struct d3d12_command_list *list,
-        VkPipelineBindPoint bind_point, unsigned int index, struct d3d12_desc *base_descriptor)
+static void vk_write_descriptor_set_for_descriptor_info(VkDescriptorSet vk_descriptor_set, uint32_t vk_binding,
+        VkDescriptorType vk_descriptor_type, union vkd3d_descriptor_info *vk_descriptor, VkWriteDescriptorSet *vk_write)
+{
+    vk_write->sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+    vk_write->pNext = NULL;
+    vk_write->dstSet = vk_descriptor_set;
+    vk_write->dstBinding = vk_binding;
+    vk_write->dstArrayElement = 0;
+    vk_write->descriptorCount = 1;
+    vk_write->descriptorType = vk_descriptor_type;
+    vk_write->pImageInfo = &vk_descriptor->image;
+    vk_write->pBufferInfo = &vk_descriptor->buffer;
+    vk_write->pTexelBufferView = &vk_descriptor->buffer_view;
+}
+
+static void d3d12_command_list_update_packed_descriptors(struct d3d12_command_list *list,
+        VkPipelineBindPoint bind_point)
 {
+    struct vkd3d_descriptor_updates *updates = &list->packed_descriptors[bind_point];
     struct vkd3d_pipeline_bindings *bindings = &list->pipeline_bindings[bind_point];
-    struct VkWriteDescriptorSet descriptor_writes[24], *current_descriptor_write;
     const struct d3d12_root_signature *root_signature = bindings->root_signature;
     const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
-    struct VkDescriptorImageInfo image_infos[24], *current_image_info;
-    const struct d3d12_root_descriptor_table *descriptor_table;
-    const struct d3d12_root_descriptor_table_range *range;
-    VkDevice vk_device = list->device->vk_device;
-    unsigned int i, j, descriptor_count;
-    struct d3d12_desc *descriptor;
+    const struct d3d12_root_descriptor_table *table;
+    VkDescriptorSet descriptor_set = VK_NULL_HANDLE;
+    union vkd3d_descriptor_info *vk_descriptor;
+    const struct d3d12_desc *base_descriptor;
+    unsigned int root_parameter_index, i, j;
+    unsigned int write_count = 0;
 
-    descriptor_table = root_signature_get_descriptor_table(root_signature, index);
+    if (!vkd3d_descriptor_updates_resize_arrays(updates, root_signature->packed_descriptor_count))
```

Comment:
I suppose for RS 1.0 we can just append new updates here and flush on submit instead of updating immediately?

Reply:
Haven't thought about that, but yeah, that might actually work.

Reply:
Actually, it's not quite as simple and would need a (rather minor) rework of how we populate the array; it still shows that this was originally written with descriptor update templates in mind. But yes, I think this is the right way forward.

Reply:
Nevermind, doesn't work. We need to pull in the correct *descriptors* at submission time, so nothing really changes from your previous attempt at implementing it.